src/jace/util/jax_helper.py

```diff
@@ -0,0 +1,201 @@
+from dataclasses import dataclass
+from typing import Any, overload
+
+import dace
+import jax.core as jax_core
+import numpy as np
+
+from jace import util
+
+
+@dataclass(init=True, repr=True, frozen=True, slots=True)
+class JaCeVar:
+    """Substitute class for Jax' `Var` instance.
+
+    This class is similar to a `jax.core.Var` class, but much simpler.
+    It is only a container for a name, shape and a datatype.
+    All extractor functions `get_jax_var{name, shape, dtype}()` will accept it, as well as multiple functions of the driver.
+
+    Notes:
+        Main intention is to test functionality.
+        If the name of a `JaCeVar` is '_' it is considered a drop variable.
+        If the name of a `JaCeVar` is empty, the automatic naming will consider it as a Jax variable.
+        The definition of `__hash__` and `__eq__` is in accordance how Jax variable works.
+    """
+
+    name: str
+    shape: tuple[int | dace.symbol | str, ...] | tuple[()]
+    dtype: dace.typeclass
+
+    def __hash__(self) -> int:
+        return id(self)
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, JaCeVar):
+            return NotImplemented
+        return id(self) == id(other)
+
+
+def get_jax_var_name(jax_var: jax_core.Atom | JaCeVar | str) -> str:
+    """Returns the name of the Jax variable as a string.
+
+    Args:
+        jax_var:     The variable to stringify.
+
+    Notes:
+        Due to some modification in Jax itself, this function is unable to return "proper" variable names.
+        This function is subject for removal.
+    """
+    match jax_var:
+        case jax_core.DropVar():
+            return "_"
+        case JaCeVar():
+            # In case of an empty name consider the jace variable as a Jax variable.
+            #  This is mostly for testing.
+            jax_name = f"jax{id(jax_var)}" if jax_var.name == "" else jax_var.name
+        case jax_core.Var():
+            # This stopped working after version 0.20.4, because of some changes in Jax
+            #  See `https://github.com/google/jax/pull/10573` for more information.
+            #  The following implementation will generate stable names, however, they will be decoupled
+            #  from output of the pretty printed Jaxpr
+            jax_name = f"jax{jax_var.count}{jax_var.suffix}"
+        case jax_core.Literal():
+            raise TypeError("Can not derive a name from a Jax Literal.")
+        case str():
+            jax_name = jax_var
+        case _:
+            raise TypeError(
+                f"Does not know how to transform '{jax_var}' (type: '{type(jax_var).__name__}') into a string."
+            )
+    assert isinstance(jax_name, str)
+
+    if not util.VALID_JAX_VAR_NAME.fullmatch(jax_name):
+        raise ValueError(f"Deduced Jax name '{jax_name}' is invalid.")
+    return jax_name
+
+
+@overload
+def get_jax_var_shape(jax_var: JaCeVar) -> tuple[int | dace.symbol | str, ...] | tuple[()]: ...
```

Comment:
What does `tuple[()]` mean in this type annotation? This is not needed for representing an empty tuple.